### PR TITLE
IO: Handle case-sensitive filesystems + cache

### LINKED ISSUE #681: Ubuntu/Linux IO issue with vita case insensitive paths
When running Vita3k on linux(in my case Ubuntu 18.04) files will fail to load due to case sensitive paths. On the Vita it's possible to name a file "app0:ImAgE.pNg" and load as "app0:image.png". I'm pretty sure that might work on windows as well... but not the case here. Below is a snippet from the vita3k log. The game is Chaos;Head Dual[PCSG00411]. This might also just be an Ext4 filesystem issue.

```
[11:47:56.285] |E| [stat_file]: Missing file at /home/user/.local/share/Vita3K/Vita3K/ux0/app/PCSG00411/USRDIR0/dual_bg.png (target path: app0:/USRDIR0/dual_bg.png)
[11:47:56.285] |W| [io_error_impl]: stat_file (sceIoGetstat) returned 0x80010002
[11:47:56.285] |E| [stat_file]: Missing file at /home/user/.local/share/Vita3K/Vita3K/ux0/app/PCSG00411/USRDIR0/dual_bg.png (target path: app0:/USRDIR0/dual_bg.png)
[11:47:56.285] |W| [io_error_impl]: stat_file (sceIoGetstat) returned 0x80010002
```
```
$ pwd
/home/user/.local/share/Vita3K/Vita3K/ux0/app/PCSG00411/USRDIR0
$ ls
DUAL_BG.png     DUAL_GAME2.png  DUAL_LOGO2.png
DUAL_GAME1.png  DUAL_LOGO1.png  DUAL_MES.png
```

[vita3k.log](https://github.com/Vita3K/Vita3K/files/4499652/vita3k.log)

### PR DESCRIPTION
**Pull request by KorewaWatchful**
Fixes #681, on case-sensitive filesystems some files couldn't be found, code attempts to iterate through game directory and find a file. 

- [x] Fix file open/stat
- [x] Fix open dir 
- [x] Support for addcont 
- [x] Lazy evaluation caching of all paths

### CODE REVIEW

File: vita3k/io/src/io.cpp [hunk 1]
```diff
@@ -244,12 +245,20 @@ SceUID open_file(IOState &io, const char *path, const int flags, const std::wstr
...
     // Do not allow any new files if they do not have a write flag.
     if (!fs::exists(system_path) && !can_write(flags)) {
-        LOG_ERROR("Missing file at {} (target path: {})", system_path.string(), path);
-        return IO_ERROR(SCE_ERROR_ERRNO_ENOENT);
+
+        const auto icase_path = device::attempt_icase_path(translated_path, system_path);
+
+        if(!icase_path.empty()) {
+            system_path = icase_path;
+        }
```
Comment (Zangetsu38) [inline vita3k/io/src/io.cpp:257]: format error here ^^

Comment (KorewaWatchful) [reply vita3k/io/src/io.cpp:257]: Lol sry, it's still WIP I'll fix everything as I move along.

**Comment by KorewaWatchful** [comment]
Ready for review, should I wrap all the stuff with `#ifdef __linux__` ?

File: vita3k/io/src/io.cpp [hunk 2]
```diff
@@ -148,6 +149,44 @@ bool init_savedata_app_path(IOState &io, const fs::path &pref_path) {
...
+void write_to_cache(IOState &io, const fs::path &system_path, const fs::path &icase_path) {
+    io.cachemap.insert(std::make_pair(system_path.string(), icase_path.string()));
+
+    fs::ofstream write_cache(io.cachefile, std::ios_base::app);
+    write_cache << system_path.string() + "," + icase_path.string() + "\n";
+    write_cache.close();
+}
+
+fs::path find_in_cache(IOState &io, const fs::path &system_path) {
+    const auto find_path = io.cachemap.find(system_path.string());
```
Comment (EXtremeExploit) [inline vita3k/io/src/io.cpp:181]: here could use std::binary_search just in case there are lots of paths

Comment (KorewaWatchful) [reply vita3k/io/src/io.cpp:181]: Requires map to be sorted XD

**Review by EXtremeExploit — DISMISSED** [review]
pog u ez pr

**Review by strikersh — CHANGES_REQUESTED** [review]
I'm against storing the cache to a file, 
because if we do we have to deal with the cache coherency.
Save file and stuff can be deleted by user or game.
For game it's easy (and should be done) just remove it from the map, but you then have to find it in the text file. If the user mess with files the cache will become incoherent

**Comment by KorewaWatchful** [comment]
Ready for review :D

**Comment by KorewaWatchful** [comment]
Would be cool if someone can test on Windows and see if the option is gone (and if everything works fine).

**Comment by EXtremeExploit** [comment]
for open dir, super meat boy uses sceIoOpenDir, you can test in that game

**Review by EXtremeExploit — APPROVED** [review]
😎

**Review by strikersh — APPROVED** [review]

**Review by Zangetsu38 — APPROVED** [review]

**Pull request merged**